ink/crates/phat_rollup_anchor_ink/src/traits/meta_transaction.rs

```diff
@@ -0,0 +1,173 @@
+pub type PrepareResult = (ForwardRequest, Hash);
+pub type MetatTxRolupCondEqMethodParams = (ForwardRequest, [u8; 65]);
+
+pub const MANAGER_ROLE: RoleType = ink::selector_id!("MANAGER_ROLE");
+
+#[derive(Debug, Eq, PartialEq, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum MetaTransactionError {
+    NonceTooLow,
+    IncorrectSignature,
+    PublicKeyNotMatch,
+    PublicKeyNotRegistered,
+    AccessControlError(AccessControlError),
+    RollupAnchorError(RollupAnchorError),
+}
+
+/// convertor from AccessControlError to MetaTxError
+impl From<AccessControlError> for MetaTransactionError {
+    fn from(error: AccessControlError) -> Self {
+        MetaTransactionError::AccessControlError(error)
+    }
+}
+
+/// convertor from RollupAnchorError to MetaTxError
+impl From<RollupAnchorError> for MetaTransactionError {
+    fn from(error: RollupAnchorError) -> Self {
+        MetaTransactionError::RollupAnchorError(error)
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub struct ForwardRequest {
+    pub from: AccountId,
+    pub nonce: Nonce,
+    pub data: Vec<u8>,
+}
+
+#[derive(Default, Debug)]
+#[openbrush::storage_item]
+pub struct Data {
+    nonces: Mapping<AccountId, Nonce>,
+    ecdsa_public_keys: Mapping<AccountId, Vec<u8>>,
+}
+
+#[openbrush::trait_definition]
+pub trait MetaTransaction:
+    Storage<Data> + EventBroadcaster + access_control::Internal + RollupAnchor
+{
+    #[ink(message)]
+    fn get_ecdsa_public_key(&self, from: AccountId) -> [u8; 33] {
+        self.data::<Data>()
+            .ecdsa_public_keys
+            .get(&from)
+            .map(|k| k.try_into().unwrap_or([0; 33]))
+            .unwrap_or([0; 33])
+    }
+
+    #[ink(message)]
+    #[openbrush::modifiers(access_control::only_role(MANAGER_ROLE))]
+    fn register_ecdsa_public_key(
+        &mut self,
+        from: AccountId,
+        ecdsa_public_key: [u8; 33],
+    ) -> Result<(), MetaTransactionError> {
+        self.data::<Data>()
+            .ecdsa_public_keys
+            .insert(&from, &ecdsa_public_key.into());
+        Ok(())
+    }
+
+    #[ink(message)]
+    fn prepare(
+        &self,
+        from: AccountId,
+        data: Vec<u8>,
+    ) -> Result<(ForwardRequest, Hash), MetaTransactionError> {
+        let nonce = self.get_nonce(from);
+
+        let request = ForwardRequest { from, nonce, data };
+        let mut hash = <Blake2x256 as HashOutput>::Type::default();
+        ink::env::hash_encoded::<Blake2x256, _>(&request, &mut hash);
+
+        Ok((request, hash.into()))
+    }
+
+    fn get_nonce(&self, from: AccountId) -> Nonce {
+        self.data::<Data>().nonces.get(&from).unwrap_or(0)
+    }
+
+    fn verify(
+        &self,
+        request: &ForwardRequest,
+        signature: &[u8; 65],
+    ) -> Result<(), MetaTransactionError> {
+        let ecdsa_public_key: [u8; 33] = self
```

Comment:
It's ok to have this version right now. However, eventually we should not use the raw signature for verification, since it doesn't have replay protection. Please consider EIP712 or similar mechanism to add domain separation.

Reply:
Not sure to fully understand the details ... I need to study a little bit more the issue, specially the EIP712.
From my understanding, if the hash is done by the phat contract and provided in the message, the issue should be fixed, right?
Can we also plan this change in the milestone 3?

Reply:
Not actually. A typical example is that if you accidentally add one public key to multiple ink contract (for verification), then the attacker can just copy the signed message to another contract.

EIP712 has a "domain separator". By default, it includes the contract address. So it prevents the above problem. The domain separator also includes some customized name and tags to further reduce the attack surface.

Reply:
Got it. I will add the contract address in the message to sign.
It might be good to implement all these changes on the EVM side as well.
The initial specification is to implement rollup for ink! based on EVM implementation.
I understand that you want to improve the current implementation and that is completely normal and I find it good (I continue to learn and the final goal is to have a very resilient version) even if it goes beyond the initial scope.
I propose to include all of these changes in the V2 expected for next milestones.

Reply:
This part is covered by the EVM implementation since it uses EIP712 for signing, instead of the plain ecdsa. Very glad to see you have already added the domain separator in this PR.